src/sdk/sql_cluster_router.cc

```diff
@@ -2651,6 +2706,67 @@ hybridse::sdk::Status SQLClusterRouter::HandleDeploy(const hybridse::node::Deplo
     sp_info.set_sql(str_stream.str());
     sp_info.set_type(::openmldb::type::ProcedureType::kReqDeployment);
 
+    // extract long windows info from select_sql
+    auto iter = deploy_node->Options()->find("long_windows");
+    std::string long_window_param = "";
+    if (iter != deploy_node->Options()->end()) {
+        long_window_param = iter->second->GetExprString();
+    }
+    std::unordered_map<std::string, std::string> long_window_map;
+    if (cluster_sdk_->IsClusterMode() && !long_window_param.empty()) {
+        if (table_pair.size() != 1) {
+            return {base::ReturnCode::kError, "unsupport multi tables with long window options"};
+        }
+        std::string base_db = table_pair.begin()->first;
+        std::string base_table = table_pair.begin()->second;
+        std::vector<std::string> windows;
+        boost::split(windows, long_window_param, boost::is_any_of(","));
+        for (auto& window : windows) {
+            std::vector<std::string> window_info;
+            boost::split(window_info, window, boost::is_any_of(":"));
+
+            if (window_info.size() == 2) {
+                long_window_map[window_info[0]] = window_info[1];
+            } else if (window_info.size() == 1) {
+                long_window_map[window_info[0]] = FLAGS_mini_window_size;
+            } else {
+                return {base::ReturnCode::kError, "illegal long window format"};
+            }
+        }
+        auto long_window_infos = base::DDLParser::ExtractLongWindowInfos(select_sql, long_window_map);
+
+        auto ns_client = cluster_sdk_->GetNsClient();
+        std::vector<::openmldb::nameserver::TableInfo> tables;
+        std::string msg;
+        ns_client->ShowTable(base_table, base_db, false, tables, msg);
+        if (tables.size() != 1) {
+            return {base::ReturnCode::kError, "table not found"};
+        }
+        std::string meta_db = openmldb::nameserver::INTERNAL_DB;
+        std::string meta_table = openmldb::nameserver::PRE_AGG_META_NAME;
+        std::string aggr_db = openmldb::nameserver::PRE_AGG_DB;
+        for (const auto& lw : long_window_infos) {
+            // insert pre-aggr meta info to meta table
+            auto aggr_table = "pre_" + deploy_node->Name() + "_" +
+                              lw.window_name_ + "_" + lw.aggr_func_ + "_" + lw.aggr_col_;
+            ::hybridse::sdk::Status status;
+            std::string insert_sql = "insert into " + meta_db + "." + meta_table + " values(\"" +
+                                     aggr_table + "\", \"" + base_db + "\", \"" + base_table + "\", \"" +
+                                     lw.aggr_func_ + "\", \"" + lw.aggr_col_ + "\", \"" + lw.partition_col_ +
+                                     "\", \"" + lw.order_col_ + "\", \"" + lw.bucket_size_ + "\");";
+            bool ok = ExecuteInsert(db, insert_sql, &status);
```

Comment:
Before creating the new pre-aggr table, we have to check whether there already exists one with the same `base_db`, `base_table`, `aggr_func`, `aggr_col`, `partition_col`, `order_col` and `bucket_size`.

If it is true, it means the pre-aggr table was already created before in some previous deployment. Then there is no need to create a new one.

Also add a test case for this scenario. 

If `base_db`, `base_table`, `aggr_func`, `aggr_col`, `partition_col`, `order_col` are the same, but `bucket_size` is different, maybe for now we just create a new one? What do you think? **Create an issue to optimize this scenario as a future work.**

Reply:
> If `base_db`, `base_table`, `aggr_func`, `aggr_col`, `partition_col`, `order_col` are the same, but `bucket_size` is different, maybe for now we just create a new one? What do you think? **Create an issue to optimize this scenario as a future work.**

This is also ok, I think it is also appropriate to bind pre-aggregated info with deployment. Compute pre-aggregated data as long as the deploys are different. But this may create redundant pre-aggregated data.

Reply:
I think we'd better try to re-use the pre-aggregated tables as much as possible.